packages/blocks/shuffle/src/shuffle.tsx

```diff
@@ -0,0 +1,153 @@
+import React, { useEffect, useRef, useState } from "react";
+import {
+  BlockComponent,
+  useGraphBlockService,
+} from "@blockprotocol/graph/react";
+// eslint-disable-next-line no-restricted-imports
+import Button from "@mui/material/Button";
+import Box from "@mui/material/Box";
+import produce from "immer";
+import { v4 as uuid } from "uuid";
+import isEqual from "lodash.isequal";
+import ShuffleIcon from "@mui/icons-material/Shuffle";
+import { ItemList } from "./components/item-list";
+
+export type Item = {
+  id: string;
+  value: string;
+};
+
+export type Items = Item[];
+
+type BlockEntityProperties = {
+  items?: string[];
+};
+
+export const initialItems = ["Thing 1", "Thing 2"];
+
+const createItems = (items: string[]) =>
+  items.map((item) => ({ id: uuid(), value: item }));
+
+export const Shuffle: BlockComponent<BlockEntityProperties> = ({
+  graph: {
+    blockEntity: {
+      entityId,
+      properties: { items },
+    },
+  },
+}) => {
+  const blockRootRef = useRef<HTMLDivElement>(null);
+  const { graphService } = useGraphBlockService(blockRootRef);
+
+  const [list, setList] = useState(
+    createItems(items?.length ? items : initialItems),
+  );
+
+  useEffect(() => {
+    if (
+      items &&
+      !isEqual(
+        items,
+        list.map((item) => item.value),
+      )
+    ) {
+      setList(createItems(items));
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [items]);
```

Comment:
Is this okay / Is there a better way to do this?

Here I want the block to update its state when `items` changes externally and is different from the current state. However I don't want for this to be triggered when `list` changes (not including it in the dependency array) because that would trigger this whenever the state changes from inside the block (breaking animations, because all items are overriden with new ids).

Reply:
Blocking:

You don't need to use an effect to update state derived from props when those props change. https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

Essentially, you should keep around a `prevItems` state, and when `items` and `prevItems` don't match, you should update `prevItems` to `items` and `list` to `createItems(items)`. You do this in the render function.